### PR TITLE
Setting for hide window when it loses focus

### LINKED ISSUE #10660: Terminal should be able to auto-minimize on focus lost
Hey!

Not sure if I missed a flag for the config somewhere, but it'd be grand if the quake mode could auto minimize itself when it loses focus. As is, I have to hit the hotkey twice to hide the window.

Thanks for the hard work.

_Originally posted by @digiwombat in https://github.com/microsoft/terminal/issues/8888#issuecomment-877944744_

### PR DESCRIPTION
**Pull request by davidegiacometti**
## Summary of the Pull Request
Added setting for for hide window when it loses focus.
Works on normal window and quake window.

 
## PR Checklist
* [x] Closes #10660
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Validation Steps Performed

- Enable **Settings > Appearance > Automatically hide window**
- Terminal window should be minimized on taskbar when it loses focus
- Quake window should be minimized on system tray when it loses focus
- Enable also **Settings > Appearance > Hide Terminal in the notification area when it is minimized**
- Terminal window should be minimized on system tray when it loses focus
- Quake window should be minimized on system tray when it loses focus

### CODE REVIEW

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 1]
```diff
@@ -442,6 +442,11 @@ long IslandWindow::_calculateTotalSize(const bool isWidth, const long clientSize
         const bool activated = LOWORD(wparam) != 0;
         _WindowActivatedHandlers(activated);
 
+        if (_isQuakeWindow && _autoHideQuakeWindow && LOWORD(wparam) == 0)
```
Comment (zadjii-msft) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:445]: This is a super nit

```suggestion
        if (_isQuakeWindow && _autoHideQuakeWindow && !activated)
```

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
Thanks for whipping this up!

This is gonna sound super annoying, but I actually envisioned this as a setting that's not only restricted to the quake window. Like, I'd bet there are plenty of folks that want their non `_quake` windows to _also_ minimize on focus lost. 

Ultimately, that would mostly mean the following changes to this PR:
1. A super tedious string replace to replace most of the `autoHideQuakeWindow` with `autoHideWindow`
2. (trickier): Right now, you just call `HideWindow`, which `SW_HIDE`s the window. For the quake window, that's fine, cause the quake window always hides on minimize. For other windows, that might not always work. We don't want to hide these windows unless `minimizeToNotificationArea` is also enabled. 


That would let this setting affect any window, regardless of whether you're in the quake window or not. Now, there's more discussion to be had over if we should opt the quake window into this behavior by default (like we do for focus mode, etc). I'm leaning no, with us relying on #9992 for enabling separate window-name-settings. But I'm also open to discussion on this one.

Does that all make sense? If not, I can try and clarify. Overall, the code looks great. It's plumbed through all the right layers, it's in the settings UI, everythink here is otherwise :chef_kiss:

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 2]
```diff
@@ -1450,4 +1449,12 @@
     <value>Bold font with bright colors</value>
     <comment>An option to choose from for the "intense text format" setting. When selected, "intense" text will be rendered as both bold text and in a brighter color</comment>
   </data>
-</root>
+  <data name="Globals_AutoHideWindow.Header" xml:space="preserve">
+    <value>Automatically hide window</value>
+    <comment>Header for a control to toggle the "Automatically hide window" setting. If enabled, Terminal window will be hidden as soon as it loses focus.</comment>
+  </data>
+  <data name="Globals_AutoHideWindow.HelpText" xml:space="preserve">
+    <value>If enabled, Terminal window will be hidden as soon as it loses focus.</value>
```
Comment (lhecker) [inline src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:1457]:
```suggestion
    <value>If enabled, the window will be hidden as soon as it loses focus.</value>
```

**Review by lhecker — APPROVED** [review]

**Review by zadjii-msft — APPROVED** [review]
easy-button.jpg

Thanks for nearing with us! This looks great to me. You're gonna make a lot of people happy 😄

**Comment by zadjii-msft** [comment]
@jvincent-siplay at the moment, no. We're tracking that in #9992.

**Comment by pas-de-2** [comment]
I don't understand the logic of why this should effect normal terminal windows. Auto hiding is completely non-standard behavior for windowed applications, while at the same time it makes little sense why a quake console should stay open on losing focus. Either option as its currently configured ends up compromising the other for standard use cases.

**Pull request merged**